src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -64,43 +74,218 @@ class ResolveExecutionDependenciesRequest:
     address: Address
     execution_dependencies: tuple[str, ...] | None
     dependencies: tuple[str, ...] | None  # can go away after 2.17.0.dev0 per deprecation
+    runnable_dependencies: tuple[str, ...] | None
 
 
 @dataclass(frozen=True)
 class ResolvedExecutionDependencies:
     digest: Digest
+    runnable_dependencies: RunnableDependencies | None
+
+
+@dataclass(frozen=True)
+class RunnableDependencies:
+    path_component: str
+    immutable_input_digests: FrozenDict[str, Digest]
+    append_only_caches: FrozenDict[str, str]
+    extra_env: FrozenDict[str, str]
+
+
+#
+# Things that need a home
+#
+
+
+@dataclass(frozen=True)
+class ExtraSandboxContents:
+    digest: Digest
+    path: str | None
+    immutable_input_digests: FrozenDict[str, Digest]
+    append_only_caches: FrozenDict[str, str]
+    extra_env: FrozenDict[str, str]
```

Comment:
I'm guilty of this myself, but now I think I'm in favour of not imposing a concrete type on the type hints, leaving it up to the implementation to pick a compatible type, such as a `FrozenDict` which may be "hidden" as an implementation detail then.

```suggestion
    immutable_input_digests: Mapping[str, Digest]
    append_only_caches: Mapping[str, str]
    extra_env: Mapping[str, str]
```

Reply:
For `dataclass(frozen=True)`, if you want the contents to truly be frozen, you need to specify a frozen member type, right? Otherwise, when you pass a `dict` in by reference, that `dict` can still be modified after the request object is created. 

```
>>> @dataclass(frozen=True)
... class A:
...   foo: Mapping[str, str]
...
>>> d = {"foo": "bar"}
>>> a = A(d)
>>> a
A(foo={'foo': 'bar'})
>>> d["baz"] = "LOLOLOLOL"
>>> a
A(foo={'foo': 'bar', 'baz': 'LOLOLOLOL'})
```

Using `FrozenDict` requires either an immutable instance be passed in, or an immutable copy be made of mutable `Mapping` implementations. This is a shortcoming of `dataclass` _viz._ `attrs` for the time being.

Reply:
As you say, dataclasses doesn't enforce types at all. You could declare your `A` with `FrozenDict` and still pass in a dict as you did. It's about intent, and `Mapping` is defined as an immutable dictionary type, there is `MutableMapping` for the regular dict type. https://docs.python.org/3/library/typing.html?highlight=mapping#typing.MutableMapping

Reply:
but, it would seem that a mutable type satisfies the smaller interface of the immutable mapping, so you're right, it's not as strictly enforced.

Reply:
The above example passes typechecking!

Per the documentation you linked `MutableMapping` is a subtype of `Mapping`, so you can pass a mutable dict in. References to that mutable dict will be treated as immutable by `mypy` (so places that receive the `dataclass` cannot mutate it without a runtime instance check), but in the original context, the object is mutable, and so is problematic for caching.

Reply:
doh, would want to be able to say `==Mapping` rather than `>=Mapping` (or `<MutableMapping`). oh well.